src/Microsoft.AspNetCore.ResponseCaching/Streams/WriteOnlySegmentStream.cs

```diff
@@ -0,0 +1,215 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.ResponseCaching.Internal
+{
+    internal class WriteOnlySegmentStream : Stream
+    {
+        private readonly List<byte[]> _segments = new List<byte[]>();
+        private readonly MemoryStream _bufferStream = new MemoryStream();
+        private readonly int _segmentSize;
+        private long _length;
+        private bool _closed;
+        private bool _disposed;
+
+        internal WriteOnlySegmentStream(int segmentSize)
+        {
+            if (segmentSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(segmentSize), segmentSize, $"{nameof(segmentSize)} must be greater than 0.");
+            }
+
+            _segmentSize = segmentSize;
+        }
+
+        // Extracting the buffered segments closes the stream for writing
+        internal List<byte[]> GetSegments()
+        {
+            if (!_closed)
+            {
+                _closed = true;
+                FinalizeSegments();
+            }
+            return _segments;
+        }
+
+        public override bool CanRead => false;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => !_closed;
+
+        public override long Length => _length;
+
+        public override long Position
+        {
+            get
+            {
+                return _length;
+            }
+            set
+            {
+                throw new NotSupportedException("The stream does not support seeking.");
+            }
+        }
+
+        private void DisposeMemoryStream()
+        {
+            // Clean up the memory stream
+            _bufferStream.SetLength(0);
+            _bufferStream.Capacity = 0;
+            _bufferStream.Dispose();
+        }
+
+        private void FinalizeSegments()
+        {
+            // Append any remaining segments
+            if (_bufferStream.Length > 0)
+            {
+                // Add the last segment
+                _segments.Add(_bufferStream.ToArray());
+            }
+
+            DisposeMemoryStream();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+
+                if (disposing)
+                {
+                    _segments.Clear();
+                    DisposeMemoryStream();
+                }
+
+                _disposed = true;
+                _closed = true;
```

Comment:
Closed means closed for writing, disposed is when the streams has been disposed segments are cleared.